rasa/core/nlg/response.py

```diff
@@ -25,28 +26,55 @@ def __init__(self, responses: Dict[Text, List[Dict[Text, Any]]]) -> None:
         """
         self.responses = responses
 
+    def _matches_filled_slots(
+        self, filled_slots: Dict[Text, Any], response: Dict[Text, Any],
+    ) -> bool:
+        """Checks if the conditional response variation matches the filled slots."""
+        constraints = response.get(RESPONSE_CONDITION)
+        for constraint in constraints:
+            name = constraint["name"]
+            value = constraint["value"]
+            if filled_slots.get(name) != value:
+                return False
+
+        return True
+
     def _responses_for_utter_action(
-        self, utter_action: Text, output_channel: Text
+        self, utter_action: Text, output_channel: Text, filled_slots: Dict[Text, Any]
     ) -> List[Dict[Text, Any]]:
-        """Return array of responses that fit the channel and action."""
-        channel_responses = []
+        """Returns array of responses that fit the channel, action and condition."""
         default_responses = []
+        conditional_responses = []
 
         for response in self.responses[utter_action]:
-            if response.get("channel") == output_channel:
-                channel_responses.append(response)
-            elif not response.get("channel"):
+            if response.get(RESPONSE_CONDITION) is None:
                 default_responses.append(response)
+            else:
+                matched_response = self._matches_filled_slots(
+                    filled_slots=filled_slots, response=response
+                )
+                if matched_response:
+                    conditional_responses.append(response)
+
+        if conditional_responses:
+            potential_responses = conditional_responses
+        else:
+            potential_responses = default_responses
+
+        channel_responses = list(
```

Comment:
I think channel filtering should happen first and then apply constraint filtering. The following scenario fails currently

```python    
    domain = Domain.from_yaml(
        """
        version: "2.0"
        responses:
           utter_action:
             - text: "example with channel"
               condition:
                - type: slot
                  name: test
                  value: A
               channel: os
             - text: "app default"
               channel: app
        """
    )
    t = TemplatedNaturalLanguageGenerator(domain.responses)
    slot = TextSlot("test", "A", influence_conversation=False)
    tracker = DialogueStateTracker(sender_id="test", slots=[slot])
    r = await t.generate("utter_action", tracker, "app")
    assert r.get("text") == "app default"
```
because the constraint matches so defaults are discarded, but then no channel + constraint option works and so None is returned.

Reply:
Reversing the order of constraints is not enough to solve for this edge case - I've just tried this different implementation, and it fails another edge case present in the unit tests (when channel is matched, but condition is not). Instead of spending too much time trying to figure this one out, I've kept the original implementation and added a new check to it and all tests now pass 🥳

Reply:
I think @kedz meant something like this?

```python
        responses_matching_channel = [
            r for r in self.responses[utter_action] if CHANNEL not in r
        ]
        if output_channel:
            responses_matching_channel = [
                response
                for response in self.responses[utter_action]
                if response.get(CHANNEL) == output_channel
            ]

        matching_both = [
            response
            for response in responses_matching_channel
            if RESPONSE_CONDITION in response
            and self._matches_filled_slots(filled_slots, response)
        ]

        if not matching_both:
            matching_both = [
                response
                for response in responses_matching_channel
                if RESPONSE_CONDITION not in response
            ]

        if matching_both:
            return matching_both

        default_answers = [
            response
            for response in self.responses[utter_action]
            if CHANNEL not in response and RESPONSE_CONDITION not in response
        ]

        return default_answers
```

Reply:
I see, I tried doing something similar but didn't succeed. 
I'm happy to keep which version is more readable 😄 No strong preference either way, although your version could end up looping twice through `self.responses[utter_action]` in the worst-case - probably not a huge concern?